### PR TITLE
dailywork-ftschindler

### PR DESCRIPTION
**Pull request by ftalbrecht**

### CODE REVIEW

File: dune/xt/grid/intersection.hh [hunk 1]
```diff
@@ -66,12 +67,18 @@ double diameter(const Intersection<G, I>& intersection)
 {
   auto max_dist = std::numeric_limits<typename G::ctype>::min();
   const auto& geometry = intersection.geometry();
-  for (auto i : Common::value_range(geometry.corners())) {
-    const auto xi = geometry.corner(i);
-    for (auto j : Common::value_range(i + 1, geometry.corners())) {
-      auto xj = geometry.corner(j);
-      xj -= xi;
-      max_dist = std::max(max_dist, xj.two_norm());
+  if (geometry.corners() == 1) { // this is the 1d case, return diameter of adjacent elements instead
```
Comment (tobiasleibner) [inline dune/xt/grid/intersection.hh]: Why is the diameter of a 1d intersection (so basically a point?) the diameter of its adjacent elements?

**Comment by ftalbrecht** [comment]
On Fri, 2019-01-25 at 15:01 -0800, Tobias Leibner wrote:
> tobiasleibner commented on this pull request.
> 
> 
> 
> > @@ -66,12 +67,18 @@ double diameter(const Intersection<G, I>&
> > intersection)
>  {
>    auto max_dist = std::numeric_limits<typename G::ctype>::min();
>    const auto& geometry = intersection.geometry();
> -  for (auto i : Common::value_range(geometry.corners())) {
> -    const auto xi = geometry.corner(i);
> -    for (auto j : Common::value_range(i + 1, geometry.corners())) {
> -      auto xj = geometry.corner(j);
> -      xj -= xi;
> -      max_dist = std::max(max_dist, xj.two_norm());
> +  if (geometry.corners() == 1) { // this is the 1d case, return
> diameter of adjacent elements instead
> 
> Why is the diameter of a 1d intersection (so basically a point?) the
> diameter of its adjacent elements?
> 

Because I am misusing this method :)

For DG methods, we require a number associated with an intersection,
for instance for the penalty term. A suitable definition is to return
the diameter of the intersection for higher dimensions, and the
diameter of the adjacent elements in one dimension.

This is probably better reflected by some free function in dune-gdt,
rather than here.

The question is what else we would like to return as diameter() in 1d
here. The old code returned 0, which may be problematic in generic code
(if you divide by this). On the other hand, returning 1 in 1d also
seems awkward.

I am open to suggestions...

-- 
 Felix Schindler

 Applied Mathematics Muenster
 University of Muenster
 Orleans-Ring 10, Room 120.012
 D-48149 Muenster

 Tel.: +49 251 83 35130
 Fax:  +49 251 83 32729
 Mail: work@felixschindler.net
 WWW:  http://felixschindler.net

**Comment by tobiasleibner** [comment]
I guess in 1d zero would be the correct return type, as the one-dimensional diameter of a point is zero. But I am also fine with returning the diameter of the adjacent elements, as long as you add a comment as why we are doing that.

**Comment by renefritze** [comment]
> This is probably better reflected by some free function in dune-gdt,
rather than here.

+1


How about we let the method return 0, which seems the mathematically correct thing to do, and emit a warning in debug mode?

**Comment by ftalbrecht** [comment]
> How about we let the method return 0, which seems the mathematically correct thing to do

Agreed.

> and emit a warning in debug mode?

Oh, I think this could well remain silent for future debugging fun :)

**Comment by ftalbrecht** [comment]
I have postponed the intersection issue for later and reverted the commit

**Comment by ftalbrecht** [comment]
merge when ready

**Pull request merged**